docs/decisions/0007-compile-common-theme-for-all-mfes.rst

```diff
@@ -0,0 +1,139 @@
+branding-specific overrides on top. They contain the bulk of the SCSS code
+generated for MFEs. Following this are usually imports for the frontend header
+and footer components, imports from the MFE own components and MFE-specific
+SCSS.
+
+This effectively means that all MFEs share a large chunk of common styling code
+with a comparatively much smaller set of styles applied over that.
+
+There is scope here to properly split the MFE stylesheet into multiple parts,
+one being the common code as represented by the above imports, and another for
+MFE-specific code, and build them separately.
+
+One thing that prevents this is that MFEs need these SCSS imports to get
+variables. In a number of cases these can be replaced with either
+Paragon/Bootstrap classes, or CSS variables. A separate ADR in `Paragon
+<https://github.com/openedx/paragon/pull/1388>`_ covers the situations where
+this isn't possible yet.
+
+Decision
+********
+
+Common cores styling that is to be applied to all MFEs should be part of a
+separate stylesheet and built separately from the MFE-specific stylesheet.
+
+Both these stylesheets will be included in each MFE, however the common
+stylesheet can be loaded from a common source for all MFEs. This will allow for
+deploying the common theme stylesheet without redeploying MFEs.
+
+The mechanism for loading this common stylesheet will need to be different from
+the mechanism for loading the MFE stylesheet. A JavaScript theme loader will
+contain the location of the SCSS file and will add this stylesheet to the
+document.
+
+The common theme package will be similar to the branding package, but can be
+deployed independently of MFEs. It will produce a CSS stylesheet, and a
+JavaScript loader that will load this stylesheet. When deploying MFEs, they can
+be provided a configuration variable that contains the location where this
+theme loader is deployed.
+
+The frontend-platform repo can host code that simplifies using themes. It can
+check if a theme loader is configured, and if so, dynamically load the theme
+from there. Otherwise, it can continue using the existing mechanism of building
+and deploying the theme for each MFE for backwards-compatibility. It can also
+provide hooks for loading the theme and potentially/eventually also selecting
+a theme.
+
+Consequences
+************
+
+With a single common stylesheet for all MFE, they should load quicker since the
+stylesheet can be cached once for all MFEs. It also makes theme rebuilds much
+quicker, and eliminates the need for rebuilding each MFE for minor change to
+the theme.
+
+For context, on a sample development machine, rebuilding an MFE takes roughly
+one minute, of this, on the same machine, around 3-4 seconds is the time it
+takes to build the theme.
+
+By continuing to support the existing theming mechanism, the current mechanism
+for theming can continue working for those who need it.
+
+Since there is a JavaScript-based loader involved, there is potential for
```

Comment:
I think that getting it right for the loader is the gist of this proposal.

Would it be possible that there is more than a loader written and which loader is used is also a configuration? I am looking at this from the lens of multitenancy probably requiring a more complex/slower loader which not everyone should pay the penalty for.

Reply:
In this proposal I imagine that the loader can be as simple as:

```JavaScript
const urlOfThemeCSS= '...';

addCSSStyleSheet(urlOfThemeCSS);
```
Where the function will essentially create a `link` element to load the stylesheet. In addition, it can have a simple API to signal to the app when the theme has loaded. 

I think a good path for multi-tenancy would be to combine this with the great work being done on runtime configuration: https://github.com/openedx/frontend-wg/issues/103

Since the loader path is a config variable, and the runtime configuration can provider URLs that vary based on the site, it would allow pointing to different themes. The loader itself can just remain simple, if that makes sense.

Reply:
What you are saying then is that the loader sis so simple that there will be no need for more than one. If the loader ends up being something so flexible, then the config variable will take care of everything.

I was trying to avoid a situation where there is already logic baked into the loader that makes it imposible to extend via configuration. It does not look like that is the case though.

Reply:
The loader ideally **needs** to be simple, since the idea is that it won't be cached. So the smaller it is the better. 

Right now the scss name is baked into the index.html file. If the stylesheet needs to load dynamically we can't do that. So we need some way to dynamically determine the name of this stylesheet file. One mechanism is to have a more complex API that returns the stylesheet location. An alternative is to have a small javascript file (the loader) that has the location of the stylesheet in it. Whenever the stylesheet location changes this file will be updated, and it will always be reloaded, so we get the latest stylesheet. 

So making the loader heavy would be detrimental to the loading speed, though there is probably nothing stopping people from doing it.

Reply:
So in the html we would have:
```html
<!DOCTYPE html>
<html>
<head>
    <script type="text/javascript" src="loaderfile_with_no_cache.js"></script>
    <script type="text/javascript" src="long_react_code.123HASH134.js"></script>
</head>
```
And if someone wants to override the loader, they can go and configure the thing to load the file from anywhere else?

e.g:
```
<!DOCTYPE html>
<html>
<head>
    <script type="text/javascript" src="different/path/custom_loaderfile_with_no_cache.js"></script>
    <script type="text/javascript" src="long_react_code.123HASH134.js"></script>
</head>
```

Reply:
@felipemontoya I think we'd load the loader itself dynamically, and just append the current timestamp as a query parameter to break the cache. 

However, the location of this loader would be a config variable provided during build, or with the runtime configuration mechanism dynamically during runtime.